tracer/src/Datadog.Trace/Debugger/Instrumentation/MethodDebuggerInvoker.MultiProbe.cs

```diff
@@ -0,0 +1,197 @@
+// <copyright file="MethodDebuggerInvoker.MultiProbe.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Debugger.Configurations.Models;
+using Datadog.Trace.Debugger.Expressions;
+using Datadog.Trace.Debugger.Instrumentation.Collections;
+using Datadog.Trace.Debugger.RateLimiting;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Instrumentation
+{
+    /// <summary>
+    /// MethodDebuggerInvoker for multiple probes scenario (where there are more than one _method_ probe).
+    /// </summary>
+    public static partial class MethodDebuggerInvoker
+    {
+        /// <summary>
+        /// Begin Method Invoker that accepts multiple probes
+        /// </summary>
+        /// <typeparam name="TTarget">Target type</typeparam>
+        /// <param name="instance">Instance value</param>
+        /// <param name="methodMetadataIndex">The index used to lookup for the <see cref="MethodMetadataInfo"/> associated with the executing method</param>
+        /// <param name="instrumentationSequence">The version of the instrumentation</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static MethodDebuggerState[] BeginMethod_StartMarker<TTarget>(TTarget instance, int methodMetadataIndex, int instrumentationSequence)
+        {
+            if (!MethodMetadataCollection.Instance.IndexExists(methodMetadataIndex))
+            {
+                Log.Warning("BeginMethod_StartMarker: Failed to receive the InstrumentedMethodInfo associated with the executing method. type = {Type}, instance type name = {Name}, methodMetadaId = {MethodMetadataIndex}, instrumentationSequence = {InstrumentationSequence}", new object[] { typeof(TTarget), instance?.GetType().Name, methodMetadataIndex, instrumentationSequence });
+                return CreateInvalidatedDebuggerStates();
+            }
+
+            ref var methodMetadataInfo = ref MethodMetadataCollection.Instance.Get(methodMetadataIndex);
+            var probesIds = methodMetadataInfo.ProbeIds;
+            var probeMetadataIndices = methodMetadataInfo.ProbeMetadataIndices;
+
+            if (instrumentationSequence != methodMetadataInfo.InstrumentationSequence)
+            {
+                Log.Warning("BeginMethod_StartMarker: The instrumentation sequence is different, received sequence: {InstrumentationSequence}, but the kept sequence is: {KeptSequenceNumber}", new object[] { instrumentationSequence, methodMetadataInfo.InstrumentationSequence });
```

Comment:
Is this something that shouldn't happen ever, or something that would happen rarely? If it's the latter, then this might be extremely problematic, because if the probe executed thousands of times, we might end up writing this warning thousands of times. 

In any case, it would be good to clarify what this invalid state means a bit, since I'm not sure I understand why we'd get here.

Reply:
It's a very rare situation where two or more instrumentation compete. To be more complete, I should handle this situation also on method exit and manage the instrumentations in a way that will allow multiple versions of the same instrumentation to co-exist without any interference.
For now, I'll get rid of it.

Do note though that for Async method (that I've just added support for) the situation is a bit difference since you have re-entrances - there facing different instrumentationVersion/Sequence is (a lot) less rare and will certainly happen.

Reply:
Done in e3443b9b41459db08459c5cbadf67dd5b0bff45c.